u8timeseries/models/fft.py

```diff
@@ -0,0 +1,125 @@
+    right after the end of 'series' in all attributes given in 'required_matches'.
+    If no such timestamp can be found, the original TimeSeries instance is returned.
+
+    :param series: TimeSeries instance to be cropped.
+    :param ts_2: Second timestamp that will be compared.
+    :required_matches: A set of pd.Timestamp attributes which will be used to choose the cropping point.
+    :return: New TimeSeries instance that is cropped as described above.
+    """
+    first_ts = series._series.index[0]
+    freq = first_ts.freq
+    pred_ts = series._series.index[-1] + freq
+
+    # start at first timestamp of given series and move forward until a matching timestamp is found
+    curr_ts = first_ts
+    while (curr_ts < pred_ts - 4 * freq):
+        curr_ts += freq
+        if compare_seasonality(pred_ts, curr_ts, required_matches):
+            new_series = series.drop_before(curr_ts)
+            return new_series
+    
+    logger.warning("No matching timestamp could be found, returning original TimeSeries.")
+    return series
+
+class FFT(AutoRegressiveModel):
+
+    def __init__(self, filter_first_n: int = -1, required_matches: set = {}, detrend: bool = False, detrend_poly_degree: int = 3):
+        """
+        Forecasting based on a discrete fourier transform using FFT of the (cropped and detrended) training sequence
+        with subsequent selection of the most significant frequencies to remove noise from the prediction.
+
+        :param filter_first_n: The total number of frequencies that will be used for forecasting.
+        :param required_matches: The attributes of pd.Timestamp that will be used to create a training
+                                 sequence that is cropped at the beginning such that the first timestamp 
+                                 of the training sequence and the first prediction point have matching 'phases'.
+                                 If the series has a yearly seasonality, include 'month', if it has a monthly 
+                                 seasonality, include 'day', etc.
+        :param detrend: Boolean value indicating whether or not detrending will be applied before performing DFT.
+        :detrend_poly_degree: The degree of the polynomial that will be used for detrending.
+
+        """
+        self.filter_first_n = filter_first_n
+        self.required_matches = required_matches
+        self.detrend = detrend
+        self.detrend_poly_degree = detrend_poly_degree
+
+    def __str__(self):
+        return 'FFT'
+
+    def chosen_frequencies(self):
+        """
+        Returns the most significant frequencies that were chosen when training the model.
+        """
+        if (not self._fit_called):
+            raise_log(Exception('fit() must be called before chosen_frequencies()'), logger)
+        frequencies = np.fft.fftfreq(len(self.fft_values))
+        chosen_frequencies = frequencies[self.filtered_indices]
+        return chosen_frequencies
+
+    @time_log(logger=logger)
```

Comment:
I wonder if it's a good idea to always log/display the time it takes. In notebooks it produces such outputs:
![image](https://user-images.githubusercontent.com/4806678/80532251-22cd3a00-899c-11ea-9761-e472e6e6e90f.png)

Reply:
kind of done, I suspended info logging for historical_forecasting